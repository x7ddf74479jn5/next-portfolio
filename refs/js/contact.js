const contactModule = (() => {
  return {
    confirmApplication: () => {
      document.getElementById("modal-switch").checked = true;
      const form = document.forms.contactForm;

      const description = convertCrlfToBr(form.description.value);
      const html = `<div id="mail-content">
                            <p>お名前：<br>${form.name.value}</p>
                            <p>メールアドレス：<br>${form.email.value}</p>
                            <p>お問い合わせ種別：<br>${form.category.value}</p>
                            <p>お問い合わせ内容：<br>${description}</p>
                          </div>`;

      document.getElementById("popup-main").innerHTML = html;
    },
    application: () => {
      document.getElementById("apply-btn").disabled = true;
      const form = document.forms.contactForm;

      if (form.email.value === "" || form.name.value === "" || form.description.value === "") {
        alert("必須項目が未入力です。");
        document.getElementById("apply-btn").disabled = false;
        return false;
      }

      // Verify email
      const regexp = /^[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;
      if (!regexp.test(form.email.value)) {
        alert("メールアドレスの形式に誤りがあります。");
        document.getElementById("apply-btn").disabled = false;
        return false;
      }

      const html = document.getElementById("mail-content").innerHTML;
      const sendMail = firebase.functions().httpsCallable("sendgridEmail");

      // Mail content for replying to the applicant
      const autoReply = {
        email: form.email.value,
        subject: "お問い合わせいただきありがとうございます【自動返信】",
        html: `<p>${form.name.value}様</p>
                       <p>
                        お問い合わせいただきありがとうございます。<br>
                        追って、当メールアドレスからご連絡いたします。
                       </p>
                       ${html}`,
      };

      // Mail content for notifying the application to admin
      const notifyApplication = {
        email: "uenot50@gmail.com",
        subject: "【自動転送】お問い合わせがありました",
        html: `<p>以下の内容でお問い合わせがありました。</p>` + html,
      };

      return Promise.all([sendMail(autoReply), sendMail(notifyApplication)])
        .then((result) => {
          console.log(result);
          alert(
            "お問い合わせありがとうございました。入力いただいたメールアドレスに自動返信メールが届いたことをご確認ください。"
          );
          return window.location.replace("/");
        })
        .catch((error) => {
          console.error(error);
          alert(
            "お問い合わせありがとうございました。入力いただいたメールアドレスに自動返信メールが届いたことをご確認ください。"
          );
          return window.location.replace("/");
        });
    },
    cancel: () => {
      document.getElementById("apply-btn").disabled = false;
      document.getElementById("modal-switch").checked = false;
    },
  };
})();
