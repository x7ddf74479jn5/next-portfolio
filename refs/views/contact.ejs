<%- include ('components/_head') %>
<%- include ('components/_header') %>


<%# content %>
<main>
  <section>
    <div class="module-spacer--medium"></div>
  </section>

  <section class="c-section">
    <div class="c-section__area-title animated fadeInLeft">
      <div class="p-headline__sub animated fadeInUp fast delay-1s">
        <h2>Contact</h2>
        <p>お問い合わせ</p>
        <ul class="u-text__breadcrumbs">
          <li><a href="/" target="_self">HOME</a><span>＞</span></li>
          <li><a href="/contact/" target="_self">CONTACT</a><span></span></li>
        </ul>
      </div>
    </div>
  </section>

  <section class="c-section">
    <div class="c-section-wrapin">
      <form class="p-forms" name="contactForm">
        <div class="p-forms__input">
          <p>
            <label>お名前 *</label>
            <input type="text" name="name" placeholder="虎 太郎" required="">
          </p>
        </div>
        <div class="p-forms__input">
          <p>
            <label>メールアドレス *</label>
            <input type="email" name="email" placeholder="your@example.com" required="">
          </p>
        </div>
        <div class="p-forms__input">
          <p>
            <label>お問い合わせ種別</label>
            <select name="category">
              <option value=""></option>
              <option value="Webサイト制作について">Webサイト制作について</option>
              <option value="Webアプリ開発について">Webアプリ開発について</option>
              <option value="自動化ツール開発について">自動化ツール開発について</option>
              <option value="Webサイト模写について">Webサイト模写について</option>
              <option value="その他">その他</option>
            </select>
          </p>
        </div>
        <div class="p-forms__input">
          <p>
            <label>お問い合わせ内容 *</label>
            <textarea name="description" rows="10" required=""></textarea>
          </p>
        </div>
        <button class="p-btn-round center" id="confirm-btn" type="button" onclick="return contactModule.confirmApplication()">内容を確認する</button>
      </form>
    </div>
  </section>
  <div class="c-modal-popup" id="popup">
    <input type="radio" name="modalPop" id="modal-switch">
    <label class="u-display-none" for="modal-switch"></label>
    <input type="radio" name="modalPop" id="close">
    <label for="close">CLOSE</label>
    <input type="radio" name="modalPop" id="pop13">
    <label for="pop13">×</label>
    <div class="c-modal-popup__body">
      <h2 class="c-modal-popup__title mb-4" id="modal-title">入力内容の確認</h2>
      <div class="c-modal-popup__main" id="popup-main">
      </div>
      <div class="p-grid__row center">
        <button class="p-btn-cancel mr-2" id="cancel-btn" type="button" onclick="return contactModule.cancel()">修正する</button>
        <button class="p-btn-apply" id="apply-btn" type="button" onclick="return contactModule.application()">送信する</button>
      </div>
    </div>
  </div>
</main>
<%# /content %>

<%- include ('components/_footer') %>