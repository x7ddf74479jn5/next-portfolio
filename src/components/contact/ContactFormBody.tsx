import type { UseFormReturn } from "react-hook-form";
import TextField from "src/components/contact/TextField";
import styles from "src/styles/components/contact/ContactForm.module.scss";

// import type { FormData } from "./ContactFormContainer";
import { Select, TextArea } from "./TextField";

export type FormData = {
  name: string;
  email: string;
  category?: string;
  description: string;
};

type Props = {
  methods: UseFormReturn<FormData>;
};

const ContactFormBody: React.VFC<Props> = ({ methods }) => {
  const {
    register,
    formState: { errors },
  } = methods;
  const options = ["", "Webサイト制作について", "Webアプリ開発について", "Webサイト模写について", "その他"];

  return (
    <>
      <div className={styles.formsInput}>
        <p>
          <label htmlFor="name">お名前 *</label>
          <TextField type="text" id="name" {...register("name")} />
        </p>
      </div>
      <div className={styles.formsInput}>
        <p>
          <label htmlFor="email">メールアドレス *</label>
          <TextField type="email" id="email" {...register("email", { required: true })} />
        </p>
        {errors.email && <span>This field is required</span>}
      </div>
      <div className={styles.formsInput}>
        <p>
          <label htmlFor="category">お問い合わせ種別</label>
          <Select id="category" {...register("category")}>
            {options.map((option, index) => {
              return (
                <option value={option} key={index}>
                  {option}
                </option>
              );
            })}
          </Select>
        </p>
      </div>
      <div className={styles.formsInput}>
        <p>
          <label htmlFor="description">お問い合わせ内容 *</label>
          <TextArea id="description" rows={10} {...register("description")} />
        </p>
      </div>
    </>
  );
};

export default ContactFormBody;
