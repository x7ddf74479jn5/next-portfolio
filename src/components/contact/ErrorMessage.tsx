import styles from "src/styles/components/contact/ContactForm.module.scss";

type Props = {
  children?: React.ReactNode;
  message?: string;
};

const ErrorMessage: React.VFC<Props> = ({ children, message }) => {
  return <p className={styles.error}>{children || message}</p>;
};

export default ErrorMessage;
