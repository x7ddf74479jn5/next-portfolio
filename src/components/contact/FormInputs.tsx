import { forwardRef } from "react";

import type { Frec } from "../../types/utils/index";

const TagNameInput = "input";
export const TextField: Frec<typeof TagNameInput> = forwardRef((props, ref) => {
  return <TagNameInput {...props} ref={ref} />;
});
TextField.displayName = "TextField";

export default TextField;

const TagNameTextArea = "textarea";
export const TextArea: Frec<typeof TagNameTextArea> = forwardRef((props, ref) => {
  return <TagNameTextArea {...props} ref={ref} />;
});
TextArea.displayName = "TextArea";

const TagNameSelect = "select";
export const Select: Frec<typeof TagNameSelect> = forwardRef((props, ref) => {
  return (
    <TagNameSelect {...props} ref={ref}>
      {props.children}
    </TagNameSelect>
  );
});
Select.displayName = "Select";
