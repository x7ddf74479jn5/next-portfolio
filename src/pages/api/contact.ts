const sendgrid = require("@sendgrid/mail");

import type { NowRequest, NowResponse } from "@vercel/node";

const SENDGRID_API_KEY = process.env.SENDGRID_API_KEY;

module.exports = async (req: NowRequest, res: NowResponse) => {
  if (req.method === "POST") {
    // Process a POST request
    sendgrid.setApiKey(SENDGRID_API_KEY);
    const message = {
      to: req.body.email,
      from: "x7ddf74479jn5@gmail.com",
      subject: req.body.subject,
      html: req.body.html,
    };
    await sendgrid.send(message);
    return res.json({ message: "ok" });
  } else {
    // Handle any other HTTP method
    res.setHeader("Allow", "POST");
    res.status(405).json({ message: `Method ${req.method} Not Allowed` });
  }
};
