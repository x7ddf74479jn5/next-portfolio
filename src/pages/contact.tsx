import { AriaTitle } from "src/components/AriaTitle";
// import { contactModule } from "src/functions/contact";
import { confirmApplication, cancel, application } from "src/lib/contact";
import { Layout } from "src/layout/layout";

const Contact: React.FC = () => {
  const handleConfirm = () => {
    confirmApplication();
  };
  const handleCancel = () => {
    cancel();
  };
  const handleApplication = () => {
    application();
  };
  // const Contact: React.FC = () => {
  //   const handleConfirm = () => {
  //     contactModule.confirmApplication();
  //   };
  //   const handleCancel = () => {
  //     contactModule.cancel();
  //   };
  //   const handleApplication = () => {
  //     contactModule.application();
  //   };

  return (
    <Layout>
      <section>
        <div className="module-spacer--medium"></div>
      </section>

      <AriaTitle title="Contact" caption="お問い合わせ" />

      <section className="c-section">
        <div className="c-section-wrapin">
          <form className="p-forms" name="contactForm">
            <div className="p-forms__input">
              <p>
                <label>お名前 *</label>
                <input type="text" name="name" placeholder="名無 太郎" required />
              </p>
            </div>
            <div className="p-forms__input">
              <p>
                <label>メールアドレス *</label>
                <input type="email" name="email" placeholder="your@example.com" required />
              </p>
            </div>
            <div className="p-forms__input">
              <p>
                <label>お問い合わせ種別</label>
                <select name="category">
                  <option value=""></option>
                  <option value="Webサイト制作について">Webサイト制作について</option>
                  <option value="Webアプリ開発について">Webアプリ開発について</option>
                  {/* <option value="自動化ツール開発について">自動化ツール開発について</option> */}
                  <option value="Webサイト模写について">Webサイト模写について</option>
                  <option value="その他">その他</option>
                </select>
              </p>
            </div>
            <div className="p-forms__input">
              <p>
                <label>お問い合わせ内容 *</label>
                <textarea name="description" rows={10} required={true}></textarea>
              </p>
            </div>
            <button className="p-btn-round center" id="confirm-btn" type="button" onClick={() => handleConfirm()}>
              内容を確認する
            </button>
          </form>
        </div>
      </section>
      <div className="c-modal-popup" id="popup">
        <input type="radio" name="modalPop" id="modal-switch" />
        <label className="u-display-none" htmlFor="modal-switch"></label>
        <input type="radio" name="modalPop" id="close" />
        <label htmlFor="close">CLOSE</label>
        <input type="radio" name="modalPop" id="pop13" />
        <label htmlFor="pop13">×</label>
        <div className="c-modal-popup__body">
          <h2 className="c-modal-popup__title mb-4" id="modal-title">
            入力内容の確認
          </h2>
          <div className="c-modal-popup__main" id="popup-main"></div>
          <div className="p-grid__row center">
            <button className="p-btn-cancel mr-2" id="cancel-btn" type="button" onClick={() => handleCancel()}>
              修正する
            </button>
            <button className="p-btn-apply" id="apply-btn" type="button" onClick={() => handleApplication()}>
              送信する
            </button>
          </div>
        </div>
      </div>
    </Layout>
  );
};

// const Contact: React.FC = () => {
//   const ()=>handleApplication() = () => {
//     console.log("return contactModule.confirmApplication()");
//     console.log("return contactModule.cancel()");
//     console.log("return contactModule.application()");
//   };

//   return (
//     <Layout>
//       <section>
//         <div className="module-spacer--medium"></div>
//       </section>

//       <AriaTitle title="Contact" caption="お問い合わせ" />

//       <section className="c-section">
//         <div className="c-section-wrapin">
//           <form className="p-forms" name="contactForm">
//             <div className="p-forms__input">
//               <p>
//                 <label>お名前 *</label>
//                 <input type="text" name="name" placeholder="名無 太郎" required />
//               </p>
//             </div>
//             <div className="p-forms__input">
//               <p>
//                 <label>メールアドレス *</label>
//                 <input type="email" name="email" placeholder="your@example.com" required />
//               </p>
//             </div>
//             <div className="p-forms__input">
//               <p>
//                 <label>お問い合わせ種別</label>
//                 <select name="category">
//                   <option value=""></option>
//                   <option value="Webサイト制作について">Webサイト制作について</option>
//                   <option value="Webアプリ開発について">Webアプリ開発について</option>
//                   {/* <option value="自動化ツール開発について">自動化ツール開発について</option> */}
//                   <option value="Webサイト模写について">Webサイト模写について</option>
//                   <option value="その他">その他</option>
//                 </select>
//               </p>
//             </div>
//             <div className="p-forms__input">
//               <p>
//                 <label>お問い合わせ内容 *</label>
//                 <textarea name="description" rows={10} required={true}></textarea>
//               </p>
//             </div>
//             <button className="p-btn-round center" id="confirm-btn" type="button" onClick={handle}>
//               内容を確認する
//             </button>
//           </form>
//         </div>
//       </section>
//       <div className="c-modal-popup" id="popup">
//         <input type="radio" name="modalPop" id="modal-switch" />
//         <label className="u-display-none" htmlFor="modal-switch"></label>
//         <input type="radio" name="modalPop" id="close" />
//         <label htmlFor="close">CLOSE</label>
//         <input type="radio" name="modalPop" id="pop13" />
//         <label htmlFor="pop13">×</label>
//         <div className="c-modal-popup__body">
//           <h2 className="c-modal-popup__title mb-4" id="modal-title">
//             入力内容の確認
//           </h2>
//           <div className="c-modal-popup__main" id="popup-main"></div>
//           <div className="p-grid__row center">
//             <button className="p-btn-cancel mr-2" id="cancel-btn" type="button" onClick={handle}>
//               修正する
//             </button>
//             <button className="p-btn-apply" id="apply-btn" type="button" onClick={handle}>
//               送信する
//             </button>
//           </div>
//         </div>
//       </div>
//     </Layout>
//   );
// };

export default Contact;
