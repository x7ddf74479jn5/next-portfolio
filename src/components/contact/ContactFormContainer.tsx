import { yupResolver } from "@hookform/resolvers/yup";
import * as React from "react";
import { SubmitHandler, useForm } from "react-hook-form";
import PrimaryButton from "src/components/common/PrimaryButton";
import ContactFormBody from "src/components/contact/ContactFormBody";
import * as yup from "yup";

import { useModalDispatch } from "../../hooks/useModalDispatch";
import { useModalState } from "../../hooks/useModalState";
import type { FormData } from "./ContactFormBody";

const schema = yup.object().shape({
  name: yup.string().required("お名前が未入力です。"),
  email: yup
    .string()
    .required("メールアドレスが未入力です。")
    .matches(
      /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/,
      "メールアドレスの形式に誤りがあります。"
    ),
  category: yup.string(),
  describe: yup.string().required("お問い合わせ内容が未入力です。"),
});
// export type FormData = yup.InferType<typeof schema>;

export const ContactFormContainer = () => {
  const methods = useForm<FormData>();
  // const methods = useForm<FormData>({ mode: "onBlur", resolver: yupResolver(schema) });
  const { handleSubmit, watch } = methods;
  const watcher = watch();

  const { isModalOpen } = useModalState();
  const { openModal, closeModal } = useModalDispatch();
  const [isFetching, setIsFetching] = React.useState(false);

  const isApplyButtonDisabled = React.useMemo(() => {
    if (isFetching) {
      return true;
    }

    return false;
  }, [isFetching]);

  const isConfirmButtonDisabled = React.useMemo(() => {
    if (isModalOpen || isFetching) {
      return true;
    }

    return false;
  }, [isModalOpen, isFetching]);

  const onSubmit = React.useCallback<SubmitHandler<FormData>>(
    (formData: FormData) => {
      openModal("CONTACT", { formData, cancel, apply, isApplyButtonDisabled });
    },
    [openModal]
  );

  const sendMessage = async (data: FormData) => {
    const url = process.env.NEXT_PUBLIC_CONTACT_API_URL;
    if (!url) return;
    return await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    });
  };

  const apply = async (data: FormData) => {
    setIsFetching(true);
    await sendMessage(data)
      .then(() => {
        alert(
          "お問い合わせありがとうございました。入力いただいたメールアドレスに自動返信メールが届いたことをご確認ください。"
        );
      })
      .catch(() => {
        alert("エラーが発生しました。恐れ入りますが再度送信してください。");
      });
    setIsFetching(false);
    closeModal();
  };

  const cancel = () => {
    closeModal();
  };

  return (
    <form className="p-forms" name="contactForm" onSubmit={handleSubmit(onSubmit)}>
      <ContactFormBody methods={methods} />

      <PrimaryButton id="confirm-btn" type="submit" disabled={isConfirmButtonDisabled}>
        内容を確認する
      </PrimaryButton>
    </form>
  );
};

export default ContactFormContainer;
