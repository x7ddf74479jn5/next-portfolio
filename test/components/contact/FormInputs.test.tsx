import type { RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Select, TextArea, TextField } from "src/components/contact/FormInputs";
import { reTestCase } from "test/test-utils";

describe("FormInput", () => {
  let renderResult: RenderResult;
  let inputElement: HTMLElement;

  describe("TextField", () => {
    beforeEach(() => {
      renderResult = render(<TextField id="id" name="name" placeholder="placeholder" type="text" />);
      inputElement = renderResult.getByRole("textbox");
    });

    it("matches snapshot", () => {
      expect(renderResult.asFragment()).toMatchSnapshot();
    });

    it("rendered correctly", () => {
      expect(inputElement).not.toHaveValue(reTestCase.anyWord);
      expect(inputElement).toHaveAttribute("id", "id");
      expect(inputElement).toHaveAttribute("name", "name");
      expect(inputElement).toHaveAttribute("placeholder", "placeholder");
      expect(inputElement).toHaveAttribute("type", "text");
    });

    it("user interaction", () => {
      userEvent.type(inputElement, "test");
      expect(inputElement).toHaveValue("test");
      userEvent.clear(inputElement);
      expect(inputElement).not.toHaveValue(reTestCase.anyWord);
    });
  });

  describe("TextArea", () => {
    beforeEach(() => {
      renderResult = render(<TextArea id="id" name="name" placeholder="placeholder" rows={10} />);
      inputElement = renderResult.getByRole("textbox");
    });

    it("matches snapshot", () => {
      expect(renderResult.asFragment()).toMatchSnapshot();
    });

    it("rendered correctly", () => {
      expect(inputElement).not.toHaveValue(reTestCase.anyWord);
      expect(inputElement).toHaveAttribute("id", "id");
      expect(inputElement).toHaveAttribute("name", "name");
      expect(inputElement).toHaveAttribute("placeholder", "placeholder");
      expect(inputElement).toHaveAttribute("rows", "10");
    });

    it("user interaction", () => {
      userEvent.type(inputElement, "test");
      expect(inputElement).toHaveValue("test");
      userEvent.clear(inputElement);
      expect(inputElement).not.toHaveValue(reTestCase.anyWord);
    });
  });

  describe("Select", () => {
    const values = ["value 1", "value 2", "value 3"];

    beforeEach(() => {
      renderResult = render(
        <Select id="id" name="name">
          {values.map((value, index) => {
            return (
              <option key={index} value={value}>
                {value}
              </option>
            );
          })}
        </Select>
      );
    });

    it("matches snapshot", () => {
      expect(renderResult.asFragment()).toMatchSnapshot();
    });

    it("rendered correctly", () => {
      const selectElement = renderResult.getByRole("combobox");
      expect(selectElement).toHaveAttribute("id", "id");
      expect(selectElement).toHaveAttribute("name", "name");
      const options = renderResult.getAllByRole("option");
      expect(options).toHaveLength(3);
      expect(options[0]).toHaveTextContent("value 1");
    });

    it("user interaction", () => {
      const selectElement = renderResult.getByRole("combobox");

      userEvent.selectOptions(selectElement, ["value 2"]);

      const option1 = renderResult.getByRole("option", { name: "value 1" }) as HTMLOptionElement;
      const option2 = renderResult.getByRole("option", { name: "value 2" }) as HTMLOptionElement;
      const option3 = renderResult.getByRole("option", { name: "value 3" }) as HTMLOptionElement;
      expect(option1.selected).toBe(false);
      expect(option2.selected).toBe(true);
      expect(option3.selected).toBe(false);
    });
  });
});
