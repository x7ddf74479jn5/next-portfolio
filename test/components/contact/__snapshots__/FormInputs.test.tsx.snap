// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput Select matches snapshot 1`] = `
<DocumentFragment>
  <select
    id="id"
    name="name"
  >
    <option
      value="value 1"
    >
      value 1
    </option>
    <option
      value="value 2"
    >
      value 2
    </option>
    <option
      value="value 3"
    >
      value 3
    </option>
  </select>
</DocumentFragment>
`;

exports[`FormInput TextArea matches snapshot 1`] = `
<DocumentFragment>
  <textarea
    id="id"
    name="name"
    placeholder="placeholder"
    rows="10"
  />
</DocumentFragment>
`;

exports[`FormInput TextField matches snapshot 1`] = `
<DocumentFragment>
  <input
    id="id"
    name="name"
    placeholder="placeholder"
    type="text"
  />
</DocumentFragment>
`;
